Local Login #210

ppelayo1 opened the pull request:
### :scroll: Description

- Describe the general shape of this PR (new feature? refactor? bug fix? one-line change?)
This pull request will implement the ability for a user to locally login to an existing account.
Used mock from https://xd.adobe.com/view/ca32af76-bb04-4e66-b0f4-0ddbb6525ea5-9f61/

- Describe what changes are being made
  Created a login portal. The portal right now only allows local login, and is not mobile responsive, but the authentication works fine. **Note** for local development must ensure line 39 of index.jsx is "uri: `http://${process.env.REACT_APP_GRAPHQL_URL}/graphql`," as http:// or https:// must be included in the uri.

  -----New -----
    **AuthRoute**
      -----Redirects to login page when user is not logged in, otherwise lets the url route complete
    **LoginHandler**
      -----Handles the logging in of the user. Performs a login query and after a successful return of the login token the App component state variable signedIn is set to true and the user is redirected to the root.
    **ThemeContext**
      -----Used to allow App Component setSignedIn and signedIn properties to be used for authentication purposes Application wide. 

  -----Modified -----
    **graphql.js**
    -----Added the loginLocal copied from the backend so that it can be used in queries
    **Index.jsx**
      -----Wrapped the routers with AuthRoute components for authentication redirects
      Checks for valid tokens and current login status thru local storage. Allows user to close tab and still remain logged in until user logs out or the token expires.
    **NavBarHeader**
    -----Added a signout button to the top, and added the code needed to perform the signout. 
      
- Describe why these changes are being made
  To provide a way to log in as an admin/user.
  I needed to implement a log in system so I can do #209

- List the use cases and edge cases relevant to this PR
  Edge case would be another site using the same localstorage keys. For this reason the keys should be something else before final deployment. Right now the keys are just "token" and "signedIn"

- Describe how errors will be handled. How will we know if this code breaks in production


- Describe any external libraries/dependencies added or removed in this PR


- Describe any security risks are there regarding this change


- Describe how you tested these changes
Log in and out. Switch tabs. Open app when token is expired.

- Link to relevant external documentation
https://www.apollographql.com/docs/react/networking/authentication/
https://xd.adobe.com/view/ca32af76-bb04-4e66-b0f4-0ddbb6525ea5-9f61/


--------------------------
### :clipboard: Mandatory Checklist

- [x] Example of a checked item (please remove when creating your Pull Request) 

- [X ] Linked to the Github Issues being addressed using the right sidebar :arrow_right:
- [X ] Have you discussed these changes with the project leader(s)?
- [X ] Do all variable and function names communicate what they do?
- [X] Were all the changes commented and / or documented?
- [X ] Is the PR the right size? (If the PR is too large to review, it might be good to break it up into multiple PRs.)
- [ ] Does all work in progress, temporary, or debugger code have a TODO comment with links to Github issues?
- [ ] *If you changed the user interface, did you add before and after screenshots to below?*

--------------------------
### :framed_picture: Screenshots and Screen Recordings

#### Recording



https://user-images.githubusercontent.com/32974267/119738978-6a8bf280-be36-11eb-9d3c-b0b98815e90c.mp4


--------------------------
### :blue_book: Glossary
- PR = Pull Request

JMStudiosJoe commented on frontend/src/components/LoginHandler/LoginHandler.jsx:
```diff
@@ -0,0 +1,90 @@
+import ThemeContext from '../ThemeContext/ThemeContext';
+
+// This function makes the query call to perform the login
+const clickHandler = (login, error, userName, password) => {
```
why is this outside the compoent?

ppelayo1 replied:
No particular reason, I can put it inside.

JMStudiosJoe replied:
ok. Thanks for clarifying. There are times when putting things outside the component will make sense but since very specific to logging in then no need.

ppelayo1 replied:
I moved the functions that were unnecessarily outside the components. LoginHandler, and NavBarHeader.

JMStudiosJoe commented on frontend/src/components/LoginHandler/LoginHandler.jsx:
```diff
@@ -0,0 +1,90 @@
+  return (
+    <div className="LoginHandler">
+      {contextType.signedIn ? <Redirect to="/" /> : ''}
+      <div id="loginHeader">CITY LOGO</div>
```
lets change these to css classes instead. Ids are very specific and should be reserved for more strict use of overriding. Please update to use className and the associated css files as well.

ppelayo1 replied:
I will do that.

ppelayo1 replied:
Using classes only now.

JMStudiosJoe commented on frontend/src/components/LoginHandler/LoginHandler.jsx:
```diff
@@ -0,0 +1,90 @@
+            <hr />
+          </div>
+
+          <form>
```
Let's get rid of the form elements, and update the input elements to just be buttons, much easier and consistent that way.

ppelayo1 replied:
I will update that.

ppelayo1 replied:
changed the form into a div, and the input tag for the button into a regular button tag.

JMStudiosJoe commented on frontend/src/components/ThemeContext/ThemeContext.jsx:
```diff
@@ -0,0 +1,6 @@
+
+// This component is used to allow the passing of the signedIn and setSignedIn props
+// to the NavBarHeader component accross the application
+const ThemeContext = React.createContext();
```
lets update the name to represent the logging in and registration instead of Theme which for UI means more of how the overall app looks and feels. This is also the first use of Context API I have seen in this app are we sure this is the best way to go about this with GraphQL? We have a method for logging in should be also just ahve a method for checking the lgin in graphql rather than adding context api?

ppelayo1 replied:
I think a graphql query to check login status would be better than using context to pass front-end state login and login setters to the navBar.

Would also need a graphql mutator to indicate that the user has been logged out.

ppelayo1 replied:
I gave the context a better name to represent its purpose. I do think having the ability to check login status thru graphql is the better option.

JMStudiosJoe reviewed (changes requested):
Looking good so far just a few questions and requesting changes.

JMStudiosJoe commented:
@anna-kartynnik could you take a look as well please?

anna-kartynnik commented:
@JMStudiosJoe sure

anna-kartynnik commented on frontend/src/components/NavBarHeader/NavBarHeader.scss:
```diff
@@ -52,11 +52,23 @@ header {
       }
+      .sign-out{
+        color: black;
+        background: #D5BBA2;
```
Here you can use color variable like `$lightBronze`.

anna-kartynnik commented on frontend/src/index.jsx:
```diff
@@ -29,11 +30,46 @@ import * as serviceWorker from './serviceWorker';
+});
+
+// this will decode a token into a usable json object, allows the access of the json properties
+const parseJwt = (token) => {
```
I'd move this to some kind of utils file. It's too specific and makes this file more difficult to read.

ppelayo1 replied:
I went ahead and did that, made two util files. verifyToken.js, and parseJwt.js.
Index.jsx is much cleaner now.

anna-kartynnik reviewed (approved):
Looks good! Nice work @ppelayo1! There are minor issues/comments/questions you can find below. One more thing regarding all the files: please use translation object for text on the pages or create a new issue where you can list all the files that require translations.

ppelayo1 commented on frontend/src/components/LoginHandler/LoginHandler.jsx:
```diff
@@ -0,0 +1,90 @@
+  const loginContext = React.useContext(LoginContext); // holds setSignIn, and signIn props
+
+  // This function makes the query call to perform the login
+  const clickHandler = () => {
```
I will definitely correct that soon. My branch to implement Google authentication has the exact problem of the handler needing to be renamed to avoid confusion.

ppelayo1 replied:
The handler is renamed to signInHandler now.

ppelayo1 commented on frontend/src/components/LoginHandler/LoginHandler.jsx:
```diff
@@ -0,0 +1,90 @@
+  const [login, { data, error }] = useLazyQuery(LOGIN_LOCAL);
+  const [userName, setUserName] = useState('');
+  const [password, setPassword] = useState('');
+  const loginContext = React.useContext(LoginContext); // holds setSignIn, and signIn props
```
I changed React.usecContext to just useContext and added useContext as an include.

anna-kartynnik commented:
Regarding errors: `error` object has `message` field that you can use to display the exact error message the backend passes to the frontend. Alternatively, you can create a new issue with the description of the problem/doubts and add a TODO comment with a link to this new issue in your code. The point is not to forget that we have a quick and easy solution in some places: if we have a specific issue about that, we won't forget.

ppelayo1 commented:
Yep I noticed that :-). Just pushed my recent changes. Need to make another push real quick. Left some literals in the navbar.

ppelayo1 commented on frontend/src/index.jsx:
```diff
@@ -29,11 +30,46 @@ import * as serviceWorker from './serviceWorker';
+//
+const authLink = setContext((_, { headers }) => {
+  // get the authentication token from local storage if it exists
+  const token = localStorage.getItem('token');
```
Just my foolishness in not being consistent. Should be consistent now. All should be window.localStorage.

ppelayo1 commented on frontend/src/index.jsx:
```diff
@@ -64,29 +100,45 @@ function SubscriptionPage() {
 }
 
+const initializeSignIn = () => {
+  const seconds = new Date() / 1000;
```
I went ahead and made a util verifyToken.js file for this to separate that functionality out.

ppelayo1 commented:
Made a number of changes.

I created some constants, and util functions that i placed in the util and constant folders to keep index more readable.
Added in the translation object and had to create some keys in the english localization folder that did not exist.
Someone that know's the language Spanish and Vietnamese will need to handle those translations. Can create a issue after this pull request is merged.
Added into the docker-compose file under environment - REACT_APP_JWT_ISSUER=ADD-ISSUER-DOMAIN. I didn't like how I was using a literal for "ADD-ISSUER-DOMAIN" of which will obviously change in the future.

ppelayo1 commented on frontend/src/components/LoginHandler/LoginHandler.scss:
```diff
@@ -0,0 +1,139 @@
+            }
+
+            .google-microsoft-login{
+                background-color:#D5BBA2;
```
The colors are using the variables now. Only two colors are not. The color for the error message on input, and the color white.

ppelayo1 commented on frontend/src/components/LoginHandler/LoginHandler.jsx:
```diff
@@ -0,0 +1,90 @@
+  useEffect(() => {
+    // Successful sign in
+    if (data) {
+      window.localStorage.setItem('token', data.loginLocal.token);
```
I went ahead and created these constants in the LocalStorageTerms.js.

ppelayo1 commented on frontend/src/components/LoginHandler/LoginHandler.jsx:
```diff
@@ -0,0 +1,90 @@
+          </div>
+
+          <div className="inputWrapper">
+            {error ? <p className="inputError">ERROR WRONG PASSWORD AND/OR EMAIL</p> : ''}
```
If password is not entered, email not entered, password wrong, email wrong, a message is displayed "ERROR WRONG PASSWORD AND/OR EMAIL"

Otherwise I just grab the "something-went-wrong" error message that already existed in the localization.

ppelayo1 commented:
also added a ErrorMessagesGraphQL constant that matches some of the error messages outputted by Apollo to the front-end.

ppelayo1 commented:
I think it is finally ready for a final review or merge. Let me know if I missed anything.
